[dynamo] support comparing stream with constant

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119199

Before the pr, we have a graph break for:
```python
def f():
    if torch.cuda.current_stream() is not None:
        return torch.randn(2, 2)
torch.compile(f, backend="eager", fullgraph=True)()
```
This pr supports comparson ops of StreamVariable and ConstantVariable by returning a constant.

It's safe to return a constant in this case becuase the StreamVariable is guarded by ID_MATCH when created.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ydwu4 committed:
[dynamo] support comparing stream with constant

[ghstack-poisoned]

Skylion007 commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1677,11 +1677,21 @@ def _unimplemented():
+            if isinstance(right, ConstantVariable) or isinstance(
+                left, ConstantVariable
+            ):
+                return ConstantVariable(op(left.value, right.value))
```
This could still fail if one of the operands does not have `.value` attr, right? That's not guaranteed to be the case here. We have a similar issue with the operator.is implementation (it caused an AttributeError).

ydwu4 replied:
I think ConstantVariable and StreamVariable are guaranteed to have `.value` attribute. So it should be safe?

Skylion007 replied:
`operator.eq is not operator.is_` on line 1688?

Skylion007 replied:
Oh yeah, I guess the invariant holds here by pidgeon hole (one operand must be ConstantVariable).

Skylion007 commented on test/dynamo/test_ctx_manager.py:
```diff
@@ -287,6 +287,24 @@ def fn(x):
+            x = torch.add(x, 2)
+
+            cur_stream = torch.cuda.current_stream()
+            if cur_stream is not None:
```
This tests for `operator.is_` `is`, the additional handling is for `operator.eq` `==` though? So do we actually have coverage here?

ydwu4 replied:
Before the PR,  we've supported operator.eq stream and stream or event and event. 

This PR adds support for comparing stream and constant. I feel in this case, the comparison logic is the same for built in comparison operators, e.g. is, is_not_, etc.

ydwu4 replied:
Added a test for `!= "const_str"`.

yifuwang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1677,11 +1677,21 @@ def _unimplemented():
+        if isinstance(left, (StreamVariable, EventVariable)) or isinstance(
+            right, (StreamVariable, EventVariable)
+        ):
+            if (
+                (isinstance(left, StreamVariable) and isinstance(right, StreamVariable))
+                or (
+                    isinstance(left, EventVariable) and isinstance(right, EventVariable)
+                )
+            ) and op is operator.eq:
```
nit: curious if this would work
```python
            if type(left) == type(right) and op is operator.eq:
```

ydwu4 replied:
This is before my change actually. I guess this would work as well and looks cleaner.

Skylion007 replied:
Better yet `type(left) is type(right)`?

ydwu4 committed:
Update on "[dynamo] support comparing stream with constant"


Before the pr, we have a graph break for:
```python
def f():
    if torch.cuda.current_stream() is not None:
        return torch.randn(2, 2)
torch.compile(f, backend="eager", fullgraph=True)()
```
This pr supports comparson ops of StreamVariable and ConstantVariable by returning a constant.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1677,11 +1677,16 @@ def _unimplemented():
+        if isinstance(left, (StreamVariable, EventVariable)) or isinstance(
+            right, (StreamVariable, EventVariable)
+        ):
+            if type(left) == type(right) and op is operator.eq:
+                return ConstantVariable(op(left.value, right.value))
```
Are we guarding on this?  I think this might be incorrect.

Can you add a test like:
```
def fn(a, s0, s1):
	if s0 == s1:
   		return a + 1
	else
		return a - 1
```

Then
1) First compile it where s0 != s1
2) Then run it again with s0 == s1 (to ensure we recompile due to the guard)

And also try the opposite of that.

ydwu4 replied:
We have a ID_MATCH guard when creating the StreamVariable. But let me add the test to make sure.

ydwu4 replied:
The added test passes. Had a discussion with @yanboliang.

ID_MATCH may not be an accurate guard. It may be too strict because we might want to re-use the compiled artifact for one stream when running in the other stream. If that's the case, we don't want to perform an ID_MATCH. But I remember the discussion, where we don't want to preserve computation stream's information since inductor will re-allocate them but not so sure about communication streams.

jansel replied:
Without an ID_MATCH we would need to guard on `s0!=s1`.

ydwu4 committed:
Update on "[dynamo] support comparing stream with constant"


Before the pr, we have a graph break for:
```python
def f():
    if torch.cuda.current_stream() is not None:
        return torch.randn(2, 2)
torch.compile(f, backend="eager", fullgraph=True)()
```
This pr supports comparson ops of StreamVariable and ConstantVariable by returning a constant.

It's safe to return a constant in this case becuase the StreamVariable is guarded by ID_MATCH when created.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
